src/metabase/driver/h2.clj

```diff
@@ -127,22 +133,57 @@
       (when (instance? SessionLocal session)
         (Parser. session)))))
 
-(defn- contains-ddl?
+(mu/defn ^:private classify-query :- [:maybe
+                                      [:map
+                                       [:command-types [:vector pos-int?]]
+                                       [:remaining-sql [:maybe :string]]]]
+  "Takes an h2 db id, and a query, returns the command-types from `query` and any remaining sql.
+   More info on command types here:
+   https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java
+
+  If the h2 parser cannot be built, returns `nil`.
+
+  - Each `command-type` corresponds to a value in org.h2.command.CommandInterface, and match the commands from `query` in order.
+  - `remaining-sql` is a nillable sql string that is unable to be classified without running preceding queries first.
+    Usually if `remaining-sql` exists we will deny the query."
   [database query]
   (when-let [h2-parser (make-h2-parser database)]
     (try
-      (let [command      (.prepareCommand h2-parser query)
-            command-type (.getCommandType command)]
-        ;; TODO: do we need to handle CommandList?
-        ;; Command types are organized with all DDL commands listed first
-        ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java
-        (< command-type CommandInterface/ALTER_SEQUENCE))
-      ;; if the query is invalid, then it isn't DDL
-      (catch Throwable _ false))))
+      (let [command            (.prepareCommand h2-parser query)
+            first-command-type (.getCommandType command)
+            command-types      (cond-> [first-command-type]
+                                 (not (instance? org.h2.command.CommandContainer command))
+                                 (into
+                                  (map #(.getType ^org.h2.command.Prepared %))
+                                  ;; when there are no fields: return no commands
+                                  (get-field command "commands" [])))]
+        {:command-types command-types
+         ;; when there is no remaining sql: return nil for remaining-sql
+         :remaining-sql (get-field command "remaining" nil)})
+      ;; only valid queries can be classified.
+      (catch org.h2.message.DbException _
+        {:command-types [] :remaining-sql nil}))))
```

Comment:
This means if a query has multiple statements and there's a syntax error in the first that means it fails to be parsed, it could mean following statements get executed, depending on whether the statements are executed independently or not. e.g. `statement with syntax error; CREATE ALIAS ...` will pass the DDL check, but the query will continue to be executed. Luckily the execution of the query is short circuited on the first syntax error and the following statements won't be executed, but that depends on how the query is executed and I'm not sure it's guaranteed to be the case. When I use the H2 web UI, a query like `statement with syntax error; CREATE ALIAS ...` will error on the first statement, and still execute the second statement.

Reply:
Interesting. That's a good thing to check. But it doesn't seem to be how the driver works in our app.

```clojure
h2=> (metabase.query-processor/process-query
       {:database 1
        :type :native
        :native {:query "select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt';"}})
Execution error (JdbcSQLSyntaxErrorException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:502).
Table "USERS" not found; SQL statement:
-- Metabase
select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt'; [42102-212]
```

Here an invalid query followed by a create alias. Our error is only about the syntax error in the first query and nothing about admin requirements in the second.

Similarly, just using the database without our ddl checks and such:

```clojure
h2=> (clojure.java.jdbc/execute!
       (metabase.driver.sql-jdbc.connection/db->pooled-connection-spec 1)
       ["select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt';"])
Execution error (JdbcSQLSyntaxErrorException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:502).
Table "USERS" not found; SQL statement:
select foo from users; CREATE ALIAS MY_SQRT for 'java.lang.Math.sqrt'; [42102-212]
h2=> (clojure.java.jdbc/execute!
       (metabase.driver.sql-jdbc.connection/db->pooled-connection-spec 1)
       ["select MY_SQRT(4)"])
Execution error (JdbcSQLSyntaxErrorException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:632).
Function "MY_SQRT" not found; SQL statement:
select MY_SQRT(4) [90022-212]
```

Trying to define the function and then use it shows it is not defined.